### ISSUE TITLE
Integer:forBy can hang interpreter with a floating point step

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#2690]
```
forBy(0, 4, 0.25, _.postln)  // hangs!!!!

forBy(0.0, 4, 0.25, _.postln)  // works correctly
```

If you actually try the first one, be ready to force-quit the interpreter.

The reason is that `Integer:forBy` uses special byte codes (inserted during class library compilation), which I guess makes the integer arithmetic faster.

But it doesn't validate the input number types. So, if `step` is a float -1.0 < step < 1.0, it's truncated to 0 and the loop never terminates.

I can pretty safely say... I thought I found some ridiculous bugs in the last year, but this one is RIDICULOUS. And it's really dangerous. Very easy to write by accident. I almost had to reboot just now.

To be honest, I think we can just delete `Integer:forBy` from the class library. This is a seldom-used method. The special byte codes do improve performance by almost a factor of 2:

```
bench { forBy(0, 1000000, 1, { nil }) }

// with Integer:forBy
time to run: 0.064958386000001 seconds.

// commenting out Integer:forBy
time to run: 0.11345979699996 seconds.
```

But the un-optimized run is still handling almost 9,000,000 iterations per second. If you're doing actual work in the loop, instead of just a no-op, the performance gain will be minimal.

If we *really* want to keep the optimization, then it *must* check the input types and fall back to the sclang implementation. But I don't think the method is used often enough to justify going to a lot of trouble.

**Comment by brianlheim** [comment]
> To be honest, I think we can just delete Integer:forBy from the class library. This is a seldom-used method.

Speak for yourself, I use this frequently.

**Comment by brianlheim** [comment]
Also hangs in this case:

```
forBy(0, 4.0, 0.25, { nil });
```

**Comment by jamshark70** [comment]
> Speak for yourself, I use this frequently.

I stand corrected :)

The user impact, however, would be a tiny, tiny fraction of a millisecond per iteration. Removing Integer:forBy would not break anything.

But I'm fine with fixing it properly, just, I won't do it myself.

**Comment by jamshark70** [comment]
https://github.com/supercollider/supercollider/blob/master/lang/LangSource/PyrInterpreter3.cpp#L1657

Quick summary of that bit:

1. If endval is a float, cast it to an integer.
2. If stepval is a float, cast it to an integer.
3. If either endval or stepval are not integers, throw an error.

*But you've already converted them* so the only protection this provides is if you passed in something that is not a simple number.

Interestingly, the original version did not cast: https://github.com/supercollider/supercollider/commit/04041aa1569d3246e44ef40c9dc890ce9c85c406#diff-4359c13e1ff2bbf03395431d8a6d6d37L1270

The cast followed about half a year later: https://github.com/supercollider/supercollider/commit/0f9b35ffe3d4d581e132622d7ec182f04f4e2fc7#diff-4359c13e1ff2bbf03395431d8a6d6d37L1297

I still think a better solution would be not to cast -- if a float is detected, fall back to the float implementation. Because:

```
forBy(0, 4, 1.25, _.postln)
0
1  // what the....?????
2
3
4
```

PS There is no warning in the documentation that the behavior is... ahem... counterintuitive if the receiver is an integer and stepval is not.

**Comment by brianlheim** [comment]
Yes absolutely, the best behavior is for this method to switch to the float implementation if any of its first three arguments is a float. Thanks so much for tracking down that code; I'm not familiar with PyrInterpreter3.cpp yet but I think this is one I can probably fix.

**Comment by brianlheim** [comment]
(but of course, it would be lovely if someone else wanted to tackle it too ^_^)

**Comment by snappizz** [comment]
I'm tagging this as high priority since it is both dangerous and entirely possible to encounter by accident.

**Comment by jamshark70** [comment]
Unfortunately (and I didn't remember this at first), we do have to be mindful of the performance optimization because:

- AppClock uses Scheduler
- `Scheduler:seconds_` uses `pairsDo`
- `pairsDo` uses `forBy`

So any performance hit would affect AppClock. Grr.

**Comment by cianoc** [comment]
Did this get fixed @brianlheim?  I'd be happy to look at this if not.

**Comment by brianlheim** [comment]
feel free @cianoc ! i never got around to it.

**Comment by lnihlen** [comment]
Not sure if folks were still working on this or no, but I made an attempt at a fix over in #3798, feedback welcome.

**Comment by snappizz** [comment]
fixed in #3804, thanks luke!
